RecoTauTag/RecoTau/interface/DeepTauHelper.h

```diff
@@ -0,0 +1,1744 @@
+                                                 deep_tau_2017::TauFunc tau_funcs) {
+    double gj_diff;
+    if (calculateGottfriedJacksonAngleDifference(tau, tau_index, gj_diff, tau_funcs))
+      return static_cast<float>(gj_diff);
+    return default_value;
+  }
+
+  bool isInEcalCrack(double eta);
+
+  template <typename Collection, typename TauCastType>
+  void fillGrids(const TauCastType& tau,
+                 const Collection& objects,
+                 deep_tau_2017::CellGrid& inner_grid,
+                 deep_tau_2017::CellGrid& outer_grid) {
+    static constexpr double outer_dR2 = 0.25;  //0.5^2
+    const double inner_radius = getInnerSignalConeRadius(tau.polarP4().pt());
+    const double inner_dR2 = std::pow(inner_radius, 2);
+
+    const auto addObject = [&](size_t n, double deta, double dphi, deep_tau_2017::CellGrid& grid) {
+      const auto& obj = objects.at(n);
+      const deep_tau_2017::CellObjectType obj_type = deep_tau_2017::GetCellObjectType(obj);
+      if (obj_type == deep_tau_2017::CellObjectType::Other)
+        return;
+      deep_tau_2017::CellIndex cell_index;
+      if (grid.tryGetCellIndex(deta, dphi, cell_index)) {
+        deep_tau_2017::Cell& cell = grid[cell_index];
+        auto iter = cell.find(obj_type);
+        if (iter != cell.end()) {
+          const auto& prev_obj = objects.at(iter->second);
+          if (obj.polarP4().pt() > prev_obj.polarP4().pt())
+            iter->second = n;
+        } else {
+          cell[obj_type] = n;
+        }
+      }
+    };
+
+    for (size_t n = 0; n < objects.size(); ++n) {
+      const auto& obj = objects.at(n);
+      const double deta = obj.polarP4().eta() - tau.polarP4().eta();
+      const double dphi = reco::deltaPhi(obj.polarP4().phi(), tau.polarP4().phi());
+      const double dR2 = std::pow(deta, 2) + std::pow(dphi, 2);
+      if (dR2 < inner_dR2)
+        addObject(n, deta, dphi, inner_grid);
+      if (dR2 < outer_dR2)
+        addObject(n, deta, dphi, outer_grid);
+    }
+  }
+
+  template <typename CandidateCastType, typename TauCastType, typename TauBlockType>
+  void createTauBlockInputs(const TauCastType& tau,
+                            const size_t& tau_index,
+                            const edm::RefToBase<reco::BaseTau> tau_ref,
+                            const reco::Vertex& pv,
+                            double rho,
+                            const deep_tau_2017::TauFunc& tau_funcs,
+                            TauBlockType& tauBlockInputs,
+                            bool disable_dxy_pca) {
+    namespace dnn = deep_tau_2017::dnn_inputs_2017_v2::TauBlockInputs;
+    namespace candFunc = deep_tau_2017::candFunc;
+
+    // TauBlockType should be a std::vector<float>::iterator (for the SonicProducer),
+    // or a tensorflow::Tensor (for the direct inference producer)
+    const auto& get = [&](int var_index) -> float& {
+      if constexpr (std::is_same_v<TauBlockType, std::vector<float>::iterator>) {
```

Comment:
in principle, this is a nice use of C++17 to avoid template metaprogramming. however, since the same construct appears 4 times, I suggest making a single function template, specialized for each case (vector or tensor), and using it in all the `create...Inputs()` functions.

Reply:
i was thinking about this at the beginning, but the shapes/dimensions for the vectors and tensors in the 4 `create...Inputs()` functions are different, and the number of parameters to determine the filling position are different (for taus only needs the `var_index`, for egamma/muon/hadrons needs the `var_index`, `idx` (tau index) and `NumberOfInputs` for different cases.)

Maybe put these into the template parameters and set the values in each `create...Inputs()` function?

Reply:
For the filling position, a shared function can just have all the possibly-necessary parameters as arguments, with default values for sometimes-unused ones.

For the tensor dimension, it may be necessary to put it into the template parameters.

Reply:
We realized that this approach would either require passing `tauBlockInputs` to every `get()` call below, or still defining a `get()` lambda that would forward `tauBlockInputs` to a common `getImpl()` function, so ultimately the duplication would not be reduced very much.